### PR TITLE
GMM UI & Cargo budget handling tweaks

### LINKED ISSUE #79115: Buying materials off the GMM without enough credits causes horrible consequences.
## Round ID:
Missing the round number but this would have occured on 10/18/23, and have confirmed locally.

## Testmerges:
N/A

## Reproduction:

Spawn in a Galactic Mineral Market.
Place an order for 500 units of any one thing. Remove credits from your ID so that you can't afford it if necessary.
Send the shuttle.

Congratuations, you have now entered a nebulous new area as follows.

The GMM prevents a single user from being able to place more than 10 stacks worth of materials in a single order. If that's the case, it will tell them they need to finish the order to buy more. However, a consequence of #78836:

> New Qol: now money will be deducted from cargo budget/your account only after the order is confirmed in the cargo request console & after shuttle is called and arrives with your packages

As these are custom orders in the same vein as mining point custom orders, cargo cannot cancel these orders, locking the user out of the GMM until they can finish their order.

An ideal fix would be for GMM orders to be cancellable, yes, but also for you to be prevented from adding more to an order that you don't reasonably have the funds in that account to afford. (Unless your paying from the cargo account, because that's not tragic, that's funny)

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
- Fixes #79115

This PR is a 2-part solution as specified by the last paragraph of the above-mentioned issue.

> but also, for you to be prevented from adding more to an order that you don't reasonably have the funds in that account to afford. 

The GMM will now not add stacks to an existing order if it exceeds the available (private or cargo depending on the mode of ordering) budget or if it exceeds the available market quantity on the market.

The UI is improved.
- Much larger to display all materials without scrolling
- Now shows the total cost of the order
- Display's number of sheets ordered next to each material type
- Has a clear button to cancel the order
- Order buttons greys out if the order quantity is greater than what's available in the market or if you don't have the available budget for it
Screenshot (340)

But this is not enough. The below example will explain why.

Say you have a budget of 500 cr in both private & cargo.
1. Place an order of materials worth 500 cr be it private or by cargo budget.
2.  Now go spend that budget on something such that it falls below 500 cr, say in this example you overspend, and the available budget becomes 0(you broke)
3. Now go call the shuttle. The order gets rejected but it's left hanging in the cargo checkout queue and will only be cleared till you raise your budget above 500 cr and again call the shuttle.

This example just applies to 500 cr but with larger amounts you order will be hanging in the queue indefinitely till your broke... makes the available budget to finally purchase it

To fix this we go to the 2nd part of the paragraph.

> An ideal fix would be for GMM orders to be cancellable.

cargo will now **"remove/cancel"** this order and will not leave it hanging in the queue indefinitely. So, this way if you by accident made a material order way beyond your available budget cargo will forgive you and clear you of that debt so you can start a fresh order.

Of course, this fix is only useful if you send the order without checking the budget but even if you do there is plan B so it's helpful to know.

## Changelog
:cl:
fix: Cargo will remove/cancel orders from its cart if that order exceeds the available budget (both private or cargo) and the player cannot cancel this order manually. All order costs are rounded up to integer values
fix: Galactic material market will deny appending stacks to your existing order if it exceeds the available (private or cargo depending on the mode of ordering) budget & if it exceeds the available materials on the market. Galactic material market UI is overall improved.
/:cl:

### CODE REVIEW

**Comment by SyncIt21** [comment]
@ArcaneMusic

File: tgui/packages/tgui/interfaces/MatMarket.tsx [hunk 1]
```diff
@@ -3,22 +3,24 @@ import { Section, Stack, Button, Modal } from '../components';
...
   quantity: number;
-  id: string; // correct this if its a number
   trend: string;
   price: number;
   color: string;
+  requested: number;
+};
+
+type Data = {
+  orderingPrive: BooleanLike; // you will need to import this
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/MatMarket.tsx:18]: can you remove this comment please lol. why is this still here

File: code/modules/cargo/materials_market.dm [hunk 2]
```diff
@@ -170,38 +236,53 @@
...
-							playsound(usr, 'sound/machines/synth_no.ogg', 35, FALSE)
-							return
-					// Append to this order
-					order.append_order(things_to_order, cost)
+			var/datum/supply_order/current_order = find_order(living_user, is_ordering_private)
+			if(!isnull(current_order))
+				// Check if this order exceeded the market limit
+				var/prior_sheets = current_order.pack.contains[sheet_to_buy]
+				if(prior_sheets + quantity > SSstock_market.materials_quantity[material_bought] )
+					say("There is not enough sheets in the market! Please wait for them to arrive before ordering more.")
```
Comment (ArcaneMusic) [inline code/modules/cargo/materials_market.dm:261]:
```suggestion
					say("There aren't enough sheets on the market! Please wait for more sheets to be traded before adding more.")
```

File: code/modules/cargo/materials_market.dm [hunk 3]
```diff
@@ -170,38 +236,53 @@
...
+				if(!isnull(current_order.paying_account)) //order is already being paid by another account
+					paying_account = current_order.paying_account
+				if(current_order.get_final_cost() + cost > paying_account.account_balance)
+					say("Order exceeds available budget!. Please send it before purchasing more.")
 					return
 
-			//Now we need to add a cargo order for quantity sheets of material_bought.sheet_type
+				// Finally Append to this order
+				current_order.append_order(things_to_order, cost)
+				say("Order appended, total cost is [current_order.get_final_cost()] cr.")
```
Comment (ArcaneMusic) [inline code/modules/cargo/materials_market.dm:282]: image


This is extremely effective at spamming the chat. I would recommend either using a non-dynamic message, using a dynamic messages and maybe a cooldown for how often they appear, or keeping the dynamic message as is and swapping to to_chat instead.

**Review by ArcaneMusic — CHANGES_REQUESTED** [review]
What I could pick out tonight, a few things

File: code/modules/cargo/materials_market.dm [hunk 4]
```diff
@@ -119,16 +172,23 @@
...
-	data["catastrophe"] = market_crashing
-	data["materials"] = material_data
-	data["creditBalance"] = balance
-	data["orderingPrive"] = ordering_private
-	data["canOrderCargo"] = can_buy_via_budget
-	return data
+	//get final order cost
+	var/current_cost = 0
+	if(!isnull(current_order))
+		current_cost = current_order.get_final_cost()
```
[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/cargo/materials_market.dm:183]: The actual cost should be adjusted in the supply order & pack level so even cargo knows about it. I'll adjust it there

[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/cargo/materials_market.dm:183]: Crate value added to `/datum/supply_pack/custom/minerals/get_cost()`

File: code/modules/cargo/materials_market.dm [hunk 5]
```diff
@@ -170,59 +242,82 @@
...
+				var/prior_sheets = current_order.pack.contains[sheet_to_buy]
+				if(prior_sheets + quantity > SSstock_market.materials_quantity[material_bought] )
+					say("There aren't enough sheets on the market! Please wait for more sheets to be traded before adding more.")
+					playsound(usr, 'sound/machines/synth_no.ogg', 35, FALSE)
+					return
+
+				// Check if the order exceeded the purchase limit
+				var/prior_stacks = ROUND_UP(prior_sheets / MAX_STACK_SIZE)
+				if(prior_stacks >= MAX_STACK_LIMIT)
+					say("There is already 10 stacks of sheets on order! Please wait for them to arrive before ordering more.")
```
Comment (Jacquerel) [inline code/modules/cargo/materials_market.dm:274]:
```suggestion
					say("There are already 10 stacks of sheets on order! Please wait for them to arrive before ordering more.")
```

File: code/modules/cargo/materials_market.dm [hunk 6]
```diff
@@ -170,59 +242,82 @@
...
+					playsound(usr, 'sound/machines/synth_no.ogg', 35, FALSE)
+					return
+
+				// Prevents you from ordering more than the available budget
+				var/datum/bank_account/paying_account = account_payable
+				if(!isnull(current_order.paying_account)) //order is already being paid by another account
+					paying_account = current_order.paying_account
+				if(current_order.get_final_cost() + cost > paying_account.account_balance)
+					say("Order exceeds available budget!. Please send it before purchasing more.")
```
Comment (Jacquerel) [inline code/modules/cargo/materials_market.dm:283]:
```suggestion
					say("Order exceeds available budget! Please send it before purchasing more.")
```

File: code/modules/cargo/materials_market.dm [hunk 7]
```diff
@@ -238,6 +333,11 @@
 	/// Is this stock block currently updating it's value with the market (aka fluid)?
 	var/fluid = FALSE
 
+/obj/item/stock_block/Initialize(mapload)
+	. = ..()
+	addtimer(CALLBACK(src, PROC_REF(value_warning)), 2.5 MINUTES)
+	addtimer(CALLBACK(src, PROC_REF(update_value)), 5 MINUTES)
```
Comment (Jacquerel) [inline code/modules/cargo/materials_market.dm:339]:
```suggestion
	addtimer(CALLBACK(src, PROC_REF(value_warning)), 2.5 MINUTES, TIMER_DELETE_ME)
	addtimer(CALLBACK(src, PROC_REF(update_value)), 5 MINUTES, TIMER_DELETE_ME)
```
Just to make sure we clean these up promptly on item destruction.

**Review by Jacquerel — CHANGES_REQUESTED** [review]
Couple of little spelling things, couple of implementation notes.

**Pull request merged**